src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,101 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.Todo;
+
+/**
+ * Handles data storage using file IO.
+ */
+public class Storage {
+    private String filePath = "./data/duke.txt";
+    private FileWriter data;
+
+    /**
+     * Constructs a new <code>Storage</code> that stores tasks from specified file.
+     *
+     * @param filePath File to be written.
+     */
+    protected Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    /**
+     * Loads, parses and returns data from specified file.
+     *
+     * @return an <code>ArrayList</code> of tasks.
+     * @throws IOException When <code>Scanner</code> does not find the file.
+     */
+    protected ArrayList<Task> load() throws IOException {
+        ArrayList<Task> arr = new ArrayList<>();
+        File f = new File(this.filePath);
+        if (!f.exists()) {
+            f.getParentFile().mkdirs();
+            f.createNewFile();
+        }
+        Scanner sc = new Scanner(f);
+        while (sc.hasNext()) {
+            String[] token = sc.nextLine().split("\\|");
+            if (token[0].equals("T")) {
+                arr.add(new Todo(token[1], Boolean.parseBoolean(token[2])));
+            } else if (token[0].equals("D")) {
+                DateTimeFormatter formatter = DateTimeFormatter.ISO_LOCAL_DATE;
+                arr.add(new Deadline(token[1], Boolean.parseBoolean(token[2]), LocalDate.parse(token[3], formatter)));
+            } else if (token[0].equals("E")) {
+                arr.add(new Event(token[1], Boolean.parseBoolean(token[2]), token[3], token[4]));
+            }
+        }
+        sc.close();
+        return arr;
+    }
+
+    /**
+     * Saves current tasks in instance to specified file.
+     *
+     * @param tl <code>ArrayList</code> of tasks to store.
+     */
+    protected void save(ArrayList<Task> tl) {
+        try {
+            this.data = new FileWriter(this.filePath);
+            tl.forEach(t -> {
+                try {
+                    if (t instanceof Todo) {
+                        this.data.write(String.format("T|%s|%s\n", t.getTaskName(), t.isDone()));
+                    } else if (t instanceof Deadline) {
+                        this.data.write(String.format("D|%s|%s|%s\n",
+                                t.getTaskName(), t.isDone(), (
+                                        (Deadline) t).getBy().format(DateTimeFormatter.ISO_LOCAL_DATE)
+                                )
+                        );
+                    } else if (t instanceof Event) {
+                        this.data.write(String.format("E|%s|%s|%s|%s\n",
+                                t.getTaskName(), t.isDone(), ((Event) t).getFrom(), ((Event) t).getTo()));
+                    }
+                } catch (IOException e) {
+                    System.out.println(e.getMessage());
+                }
+            });
```

Comment:
Wow this is really difficult to read, the level of indentation seems to be off and the brackets are not leveled properly! May I suggest reconfiguring your wrapping settings in your IDE auto formatter if you are using one. Also nits tl could be better phrased into its full name for better readability in this case to differentiate between t of task

Reply:
Thanks for your suggestion, I think I was merely following the checkstyle back then. If you don't mind me asking, where would the text wrap setting be in the IDE for this particular case? Much appreciated

Reply:
If you are using IntelliJ -> settings -> code style -> java -> wrapping and braces. You can tweak the settings there to your preference!

Reply:
Awesome, thanks a lot!